[ht16k33_char] Add support for HT16K33 based character displays

ilikecake opened the pull request:
# What does this implement/fix?

Adds support for HT16K33 based character displays like [this one](https://www.adafruit.com/product/1911). Based on the [HT16k33_alpha](https://github.com/ssieb/esphome_components/tree/master/components/ht16k33_alpha) library from @ssieb, but generalized to support other devices.

Note for anyone that wants to use this component: I have split the code for this into several PRs to make it easier to review. Until all the PRs are accepted, you will want to use the external component in my repo that contains all the functionality. To do that, add the below to your YAML

```
external_components:
  - source: github://ilikecake/esphome-external
    refresh: 1h
```

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Requires: https://github.com/esphome/esphome/pull/8454

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4771

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
display:
  - platform: ht16k33_char
    device: ADAFRUIT_7_SEG_.56IN
    address: 0x70
    max_buffer_length: 8
    lambda: |-
      it.print(0, true, "1234");

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

ilikecake committed:
Add a function to return the I2C address

ilikecake committed:
Merge branch 'dev' into add_i2c_get_address

ilikecake committed:
Make the get address function const.

Co-authored-by: Djordje Mandic <6750655+DjordjeMandic@users.noreply.github.com>

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Move CONF_CONTINUOUS to const.py

ilikecake committed:
Move files over to the correct repository

ilikecake committed:
Update CODEOWNERS

ilikecake committed:
Change the brightness function to allow setting 0-16.

ilikecake committed:
Add tests

ilikecake committed:
Fix formatting

jesserockz committed:
Merge branch 'dev' into ht16k33-char

jesserockz commented on esphome/components/esp32_ble_tracker/__init__.py:
> Create a separate PR if/when you wish to move a constant into esphome/const.py.

https://developers.esphome.io/contributing/code/#esphome-specific-idiosyncrasies

ilikecake replied:
Thanks for the review. I have created a PR to update cost.py with the new variable. Please let me know if you have any other comments/suggestions on this code.

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Revert changes made to other components

These changes were implemented in a separate PR. They are not needed here.

ilikecake committed:
Merge branch 'ht16k33-char' of https://github.com/ilikecake/esphome into ht16k33-char

ilikecake committed:
Add support for Sparkfun 14 Segment devices.

ilikecake committed:
Formatting fixes

ilikecake committed:
Fix formatting.

ilikecake committed:
Fix formatting.

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Switch to precomputed font tables

ilikecake committed:
Merge branch 'ht16k33-char' of https://github.com/ilikecake/esphome into ht16k33-char

ilikecake committed:
Correct font conversion errors

ilikecake committed:
Formatting fix.

ilikecake committed:
Merge branch 'dev' into ht16k33-char

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake commented:
There is not really a way to split this up. If I want to support multiple devices, I need separate font lookup tables for each one. If you have any suggestions on how I can split this up, I am all ears.

ilikecake commented:
> Thank you! Displays are already on the way, I will test your component as soon as they arrive.

Thanks for testing this. Please let me know if you have any issues.

B48D81EFCC commented:
> Thanks for testing this. Please let me know if you have any issues.
Unfortunately, the one I ordered (and arrived already) seems to be broken. I can't establish a I2C connection.
I ordered a few more, from different suppliers/vendors and in different colors. But it will take 1 or 2 weeks until they arrive.

Maybe a bit off-topic: I tested successfully the LED matrix version of this display/IC: https://github.com/esphome/esphome/pull/6448

Is there any plan to incorporate the matrix functionality in your component or something like that?

ilikecake commented:
> Maybe a bit off-topic: I tested successfully the LED matrix version of this display/IC: #6448 Is there any plan to incorporate the matrix functionality in your component or something like that?

I thought about that, but figured that there would not be much commonality between character and matrix displays. The way you implement the lambda functions is probably completely different between the two. I figured keeping them as separate devices would make things easier (hence why I called this one 'ht16k33-char'). If you have ideas on how they could be combined, let me know.

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

B48D81EFCC commented:
Hi,

I would like to print temperatures in degree Celsius. E.g: 23°C
Is there any way to print the ° character? Using the upper part of a 7 segment digit would work I think.

ilikecake commented:
> Hi,
> 
> I would like to print temperatures in degrees Celsius, e.g., 23 °C. Is there any way to print the `°` character? Using the upper part of a 7-segment digit might work, I think.
> 
> Another way would be to use the FLIPPED variant and use the decimal point. ~But the following seems not to work:~ `it.print(0, true, "23.C");`
> 
> ~The decimal dot is omitted.~
> 
> After reading your documentation, I learned that you have to use `'` instead of `.` when using the FLIPPED version: `it.print(0, true, "23'C");`
> 
> Regarding the implementation of `°`, I might create a new PR to support this in future, since this PR is already large.

Thanks for testing this. I hadn't considered displaying a degree symbol on the 7-segment display. The allowable characters are stored in the char_map for the device. I am wondering if there would be a way to make an `add_char_to_map()` function or something that would let people add custom characters to the map. I will think about this a bit and see if I can figure out how to implement it.

As you say this PR is quite large, so maybe something that can be added later?

B48D81EFCC commented:
Hi, I also tested this PR with a 14-segment device and it worked fine.

Are there any differences in implementation between the different vendors, other than the number of segments (7 vs. 14)?
They all use the HT16K33 IC.

B48D81EFCC commented:
I the following result correct, based on this config?:
``` yaml
display:
  - platform: ht16k33_char
    device: ADAFRUIT_14_SEG_FLIPPED
    address: 0x70
    buffer_size: 16
    scroll: false
    brightness: 16
    lambda: |-
      it.printf(0, true, "21'C");
```

<img width="485" height="248" alt="image" src="https://github.com/user-attachments/assets/9509b72d-7029-40c0-8e1d-a8d3d1b25d2f" />

Compared to the 7-segment device:
``` yaml
display:
  - platform: ht16k33_char
    device: ADAFRUIT_7SEGMENT_.56IN_FLIPPED
    address: 0x70
    buffer_size: 16
    scroll: false
    brightness: 16
    lambda: |-
      it.printf(0, true, "21'C");
```

<img width="530" height="216" alt="IMG_2573" src="https://github.com/user-attachments/assets/08c19419-648c-4e6a-aa85-1124193defc7" />

ilikecake commented:
> Is the following result correct, based on this config?:

Yes. Because the 14 segment display has `'` as one of its valid characters, it displays that instead of lighting the decimal point.

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Move the character map to the struct and allow modification

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake commented:
> Hi,
> 
> I would like to print temperatures in degrees Celsius, e.g., 23 °C. Is there any way to print the `°` character? Using the upper part of a 7-segment digit might work, I think.
> 
> Another way would be to use the FLIPPED variant and use the decimal point. ~But the following seems not to work:~ `it.print(0, true, "23.C");`
> 
> ~The decimal dot is omitted.~
> 
> After reading your documentation, I learned that you have to use `'` instead of `.` when using the FLIPPED version: `it.print(0, true, "23'C");`
> 
> Regarding the implementation of `°`, I might create a new PR to support this in future, since this PR is already large.

Well, I couldn't leave it enough alone :P 

I rearranged how I handled the font table and added a `remove_characters: ` and `add_characters: ` config options to the yaml. It still needs some error checking, but it works with all my test devices. I need to add to the documentation giving instructions on how it works, but if you want to test it, let me know and I can give a preview.

ilikecake committed:
precommit fixes

ilikecake committed:
Merge branch 'ht16k33-char' of https://github.com/ilikecake/esphome into ht16k33-char

ilikecake committed:
Update to the new i2c write function format.

ilikecake committed:
Update i2c write commands to new format.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Rename buffer to message_buffer_

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

B48D81EFCC commented:
Hi @ilikecake,

awesome. I was able to add the `°` character.
But it looks like, that no character is printed after the custom character was printed.

I have the following config, with the following result:
```yaml
display:
  - platform: ht16k33_char
    device: ADAFRUIT_7SEGMENT_.56IN_FLIPPED
    address: 0x70
    lambda: |-
      it.printf(0, true, "21°C");
    update_interval: 5s
    add_characters: { "°": 0b001000001100011 }
```

 I expected 21°C.

<img width="565" height="237" alt="IMG_2747" src="https://github.com/user-attachments/assets/204b07f2-df71-451c-842d-7b17080fb214" />

ilikecake commented:
> I expected 21°C.

I am able to reproduce this with my devices. This appears to be happening because the '°' symbol is represented by a two byte char, and I did not expect that in my code. I will try to think of a way to handle this. In the mean time, if you want to cheat, you can assign the degree symbol to a 'normal' character that the display does not support (like 'M', for example). You could then tell it to print '21MC' and it will translate the M into the character you want.

ilikecake committed:
Rework code to make a common send_to_device function.
Rework code to allow for UTF-8 multi-byte characters.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Update test with correct parameter name.

ilikecake committed:
Change from millis() to App.get_loop_component_start_time()

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

B48D81EFCC commented:
Hey @ilikecake,

your workaround (replacing an existing character) worked. Thanks

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake commented:
> Hey @ilikecake,
> 
> your workaround (replacing an existing character) worked. Thanks

Thanks for verifying. The latest updates *should* allow for up to 4-byte unicode characters so you won't need that workaround. However, my dev environment stopped working so I need to figure out how to fix that before I can clean up the code and verify.

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Merge branch 'esphome:dev' into ht16k33-char

ilikecake committed:
Add config validation to add/remove characters

ilikecake committed:
Simplify send to display function

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Updates for clang-tidy

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Remove extra comments

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Cleanup

ilikecake committed:
Merge branch 'ht16k33-char' of https://github.com/ilikecake/esphome into ht16k33-char

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Make number of character per display settable

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Mask invalid parts of the character code.

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Remove some devices to make it easier to review.

ilikecake committed:
Removed devices to make the PR smaller.

ilikecake committed:
Re-add other devices

ilikecake commented:
> What happened to the Adafruit 7 segment display support?

Sorry, in an effort to reduce the code size and get this reviewed, I removed some of the devices. I have added them back in, so you should be able to use them again. TBH, it probably doesn't matter, I was not able to reduce the code size very much.

To any reviewers that might be reading this, if you would like me to remove the other devices again to reduce the code size from 1690 to 1193, I can do that. However, I would rather not do that unless a review is imminent so as to not break the PR for anyone that might be using it.

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake commented:
I can give this a try.

I tried removing all but the 7 segment device, but it was still a bit over the cutoff of 1000 lines to give it the tag `too-big`. There were other people using the other devices on this PR, so I didn't want to disrupt their usage by removing large sections of the code for an unknown amount of time. I guess with the external component updated, anyone that wants to use this component can use that until all the parts of this get merged.

ilikecake committed:
Remove functionality to try to get the PR below 1000 lines

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Remove other scrolling related functions

ilikecake committed:
Fix tests.

ilikecake committed:
Merge branch 'ht16k33-char' of https://github.com/ilikecake/esphome into ht16k33-char

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Remove unused variables

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Modernize namespace callouts

ilikecake committed:
Merge branch 'ht16k33-char' of https://github.com/ilikecake/esphome into ht16k33-char

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char

ilikecake committed:
Merge branch 'dev' into ht16k33-char
